### PR TITLE
emu_file. refactor code for safty and compatibility.

### PR DESCRIPTION
**Pull request by weia5-emc**
1. keep layout of bin file unchanged.
2. save to file after data is ready.

### CODE REVIEW

**Comment by emcinfrasim** [comment]
<details><summary>Test Error Details</summary><p>
TestName: test_fru_pn_sn 
<br>ErrorType: exceptions.AssertionError 
<br>ErrorMessage: line None, Failed to get pn from node 192.168.188.91
-------------------- >> begin captured logging << --------------------
ivn.core: INFO: [Validate topology]
ivn.core: INFO: All namespaces in index have definition.
ivn.core: INFO: All openvswitch in index have definition.
ivn.core: INFO: Namespace node0ns interfaces and bridges config is legal.
ivn.core: INFO: Namespace node1ns interfaces and bridges config is legal.
ivn.core: INFO: Openvswitch br-int ports config is legal.
ivn.core: INFO: Connection from namespace interface ns0-einf0 to openvswitch port vint0 is defined.
ivn.core: INFO: Connection from namespace interface ns1-einf0 to openvswitch port vint1 is defined.
ivn.core: INFO: Topology pass validation.
ivn.core: INFO: [Define openvswitches]
ivn.core: INFO: vswitch br-int is created.
ivn.core: INFO: port vint0 is added to br-int.
ivn.core: INFO: port vint1 is added to br-int.
ivn.core: INFO: Openvswitch br-int is defined in /etc/network/interfaces.d and restarted.
ivn.core: INFO: [Define namespaces]
ivn.core: INFO: namespace node1ns is created.
ivn.core: INFO: Setting Id 1 for namespace node1ns.
ivn.core: INFO: Validation of creating namespace node1ns passed.
ivn.core: INFO: interface ns1-einf0 in namespace node1ns is created, peer: vint1.
ivn.core: INFO: bridge br0 in namespace node1ns is created on interface ns1-einf0.
ivn.core: INFO: namespace node0ns is created.
ivn.core: INFO: Setting Id 2 for namespace node0ns.
ivn.core: INFO: Validation of creating namespace node0ns passed.
ivn.core: INFO: interface ns0-einf0 in namespace node0ns is created, peer: vint0.
ivn.core: INFO: bridge br0 in namespace node0ns is created on interface ns0-einf0.
ivn.core: INFO: [Set openvswitch ports up]
ivn.core: INFO: set openvswitch br-int up.
ivn.core: INFO: set port vint0 up.
ivn.core: INFO: set port vint1 up.
ivn.core: INFO: [Set namespace interfaces up]
ivn.core: INFO: namespace node1ns interfaces are defined in /etc/netns/node1ns/network.
ivn.core: INFO: interfaces in namespace node1ns are restarted.
ivn.core: INFO: namespace node0ns interfaces are defined in /etc/netns/node0ns/network.
ivn.core: INFO: interfaces in namespace node0ns are restarted.
ivn.core: INFO: [Setup portforward]
ivn.core: INFO: forwarding from 192.168.188.91:5901 to host:15901
ivn.core: INFO: forwarding from 192.168.188.92:5901 to host:25901
ivn.core: INFO: forwarding from 192.168.188.91:8022 to host:18022
ivn.core: INFO: forwarding from 192.168.188.92:8022 to host:28022
4Model: INFO: [compute] infrasim can't find numactl in this environment
4Model: INFO: [BaseDrive] Creating drive: /tmp/sas_0.img
4Model: INFO: [BaseDrive] Creating drive: /tmp/disk-nvme-3.img
5Model: INFO: [compute] infrasim can't find numactl in this environment
chassis_test: INFO: chassis_test-chassis's command line: infrasim-chassis chassis_test /home/infrasim/.infrasim/chassis_test/data/shm_data.bin
chassis_test: INFO: [ 23336  ] chassis_test-chassis starts to run
4Model: INFO: chassis_test_node_0-socat's command line: /usr/bin/socat pty,link=/home/infrasim/.infrasim/chassis_test_node_0/.pty0_chassis_test_node_0,waitslave,unlink-close=0 unix-listen:/home/infrasim/.infrasim/chassis_test_node_0/.serial,fork
4Model: INFO: [ 23355  ] chassis_test_node_0-socat starts to run
4Model: INFO: chassis_test_node_0-bmc's command line: ipmi_sim -c /home/infrasim/.infrasim/chassis_test_node_0/etc/vbmc.conf -f /home/infrasim/.infrasim/chassis_test_node_0/data/dell_r730.emu -n -s /home/infrasim/.infrasim/chassis_test_node_0/data -l /var/log/infrasim/chassis_test_node_0/ipmi_sim.log -m share_mem_chassis_test
4Model: INFO: [ 23358  ] chassis_test_node_0-bmc starts to run
4Model: INFO: [ 23368  ] chassis_test_node_0-node is running
4Model: INFO: chassis_test_node_0-racadm's command line: racadmsim chassis_test_node_0 0.0.0.0 10022 admin admin auto
4Model: INFO: [ 23384  ] chassis_test_node_0-racadm starts to run
4Model: INFO: chassis_test_node_0-monitor's command line: infrasim-monitor chassis_test_node_0 0.0.0.0 9005
4Model: INFO: [ 23388  ] chassis_test_node_0-monitor starts to run
5Model: INFO: chassis_test_node_1-socat's command line: /usr/bin/socat pty,link=/home/infrasim/.infrasim/chassis_test_node_1/.pty0_chassis_test_node_1,waitslave,unlink-close=0 unix-listen:/home/infrasim/.infrasim/chassis_test_node_1/.serial,fork
5Model: INFO: [ 23392  ] chassis_test_node_1-socat starts to run
5Model: INFO: chassis_test_node_1-bmc's command line: ipmi_sim -c /home/infrasim/.infrasim/chassis_test_node_1/etc/vbmc.conf -f /home/infrasim/.infrasim/chassis_test_node_1/data/dell_r730.emu -n -s /home/infrasim/.infrasim/chassis_test_node_1/data -l /var/log/infrasim/chassis_test_node_1/ipmi_sim.log -m share_mem_chassis_test
5Model: INFO: [ 23398  ] chassis_test_node_1-bmc starts to run
5Model: INFO: [ 23408  ] chassis_test_node_1-node is running
5Model: INFO: chassis_test_node_1-racadm's command line: racadmsim chassis_test_node_1 0.0.0.0 10022 admin admin auto
5Model: INFO: [ 23421  ] chassis_test_node_1-racadm starts to run
5Model: INFO: chassis_test_node_1-monitor's command line: infrasim-monitor chassis_test_node_1 0.0.0.0 9005
5Model: INFO: [ 23425  ] chassis_test_node_1-monitor starts to run
paramiko.transport: DEBUG: starting thread (client mode): 0x5c44f050L
paramiko.transport: DEBUG: Local version/idstring: SSH-2.0-paramiko_2.4.2
paramiko.transport: ERROR: Exception: Error reading SSH protocol banner
paramiko.transport: ERROR: Traceback (most recent call last):
paramiko.transport: ERROR:   File "/home/infrasim/jenkins/workspace/Infrasim_Compute_pull_request_trigger/infrasim-compute/.tox/functional/local/lib/python2.7/site-packages/paramiko/transport.py", line 1966, in run
paramiko.transport: ERROR:     self._check_banner()
paramiko.transport: ERROR:   File "/home/infrasim/jenkins/workspace/Infrasim_Compute_pull_request_trigger/infrasim-compute/.tox/functional/local/lib/python2.7/site-packages/paramiko/transport.py", line 2143, in _check_banner
paramiko.transport: ERROR:     "Error reading SSH protocol banner" + str(e)
paramiko.transport: ERROR: SSHException: Error reading SSH protocol banner
paramiko.transport: ERROR: 
paramiko.transport: DEBUG: starting thread (client mode): 0x5c4f6050L
paramiko.transport: DEBUG: Local version/idstring: SSH-2.0-paramiko_2.4.2
paramiko.transport: DEBUG: Remote version/idstring: SSH-2.0-OpenSSH_7.6p1 Ubuntu-4
paramiko.transport: INFO: Connected (version 2.0, client OpenSSH_7.6p1)
paramiko.transport: DEBUG: kex algos:[u'curve25519-sha256', u'curve25519-sha256@libssh.org', u'ecdh-sha2-nistp256', u'ecdh-sha2-nistp384', u'ecdh-sha2-nistp521', u'diffie-hellman-group-exchange-sha256', u'diffie-hellman-group16-sha512', u'diffie-hellman-group18-sha512', u'diffie-hellman-group14-sha256', u'diffie-hellman-group14-sha1'] server key:[u'ssh-rsa', u'rsa-sha2-512', u'rsa-sha2-256', u'ecdsa-sha2-nistp256', u'ssh-ed25519'] client encrypt:[u'chacha20-poly1305@openssh.com', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr', u'aes128-gcm@openssh.com', u'aes256-gcm@openssh.com'] server encrypt:[u'chacha20-poly1305@openssh.com', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr', u'aes128-gcm@openssh.com', u'aes256-gcm@openssh.com'] client mac:[u'umac-64-etm@openssh.com', u'umac-128-etm@openssh.com', u'hmac-sha2-256-etm@openssh.com', u'hmac-sha2-512-etm@openssh.com', u'hmac-sha1-etm@openssh.com', u'umac-64@openssh.com', u'umac-128@openssh.com', u'hmac-sha2-256', u'hmac-sha2-512', u'hmac-sha1'] server mac:[u'umac-64-etm@openssh.com', u'umac-128-etm@openssh.com', u'hmac-sha2-256-etm@openssh.com', u'hmac-sha2-512-etm@openssh.com', u'hmac-sha1-etm@openssh.com', u'umac-64@openssh.com', u'umac-128@openssh.com', u'hmac-sha2-256', u'hmac-sha2-512', u'hmac-sha1'] client compress:[u'none', u'zlib@openssh.com'] server compress:[u'none', u'zlib@openssh.com'] client lang:[u''] server lang:[u''] kex follows?False
paramiko.transport: DEBUG: Kex agreed: ecdh-sha2-nistp256
paramiko.transport: DEBUG: HostKey agreed: ssh-ed25519
paramiko.transport: DEBUG: Cipher agreed: aes128-ctr
paramiko.transport: DEBUG: MAC agreed: hmac-sha2-256
paramiko.transport: DEBUG: Compression agreed: none
paramiko.transport: DEBUG: kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
paramiko.transport: DEBUG: Switch to new keys ...
paramiko.transport: DEBUG: Adding ssh-ed25519 host key for [192.168.188.91]:8022: c857a7912437d0104fb5282bbb400087
paramiko.transport: DEBUG: Trying discovered key 6fdbaea68289971a2e37c6caca1c87f6 in /home/infrasim/.ssh/id_rsa
paramiko.transport: DEBUG: userauth is OK
paramiko.transport: INFO: Authentication (publickey) failed.
paramiko.transport: DEBUG: userauth is OK
paramiko.transport: INFO: Authentication (password) successful!
paramiko.transport: DEBUG: starting thread (client mode): 0x5e3bd3d0L
paramiko.transport: DEBUG: Local version/idstring: SSH-2.0-paramiko_2.4.2
paramiko.transport: DEBUG: Remote version/idstring: SSH-2.0-OpenSSH_7.6p1 Ubuntu-4
paramiko.transport: INFO: Connected (version 2.0, client OpenSSH_7.6p1)
paramiko.transport: DEBUG: kex algos:[u'curve25519-sha256', u'curve25519-sha256@libssh.org', u'ecdh-sha2-nistp256', u'ecdh-sha2-nistp384', u'ecdh-sha2-nistp521', u'diffie-hellman-group-exchange-sha256', u'diffie-hellman-group16-sha512', u'diffie-hellman-group18-sha512', u'diffie-hellman-group14-sha256', u'diffie-hellman-group14-sha1'] server key:[u'ssh-rsa', u'rsa-sha2-512', u'rsa-sha2-256', u'ecdsa-sha2-nistp256', u'ssh-ed25519'] client encrypt:[u'chacha20-poly1305@openssh.com', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr', u'aes128-gcm@openssh.com', u'aes256-gcm@openssh.com'] server encrypt:[u'chacha20-poly1305@openssh.com', u'aes128-ctr', u'aes192-ctr', u'aes256-ctr', u'aes128-gcm@openssh.com', u'aes256-gcm@openssh.com'] client mac:[u'umac-64-etm@openssh.com', u'umac-128-etm@openssh.com', u'hmac-sha2-256-etm@openssh.com', u'hmac-sha2-512-etm@openssh.com', u'hmac-sha1-etm@openssh.com', u'umac-64@openssh.com', u'umac-128@openssh.com', u'hmac-sha2-256', u'hmac-sha2-512', u'hmac-sha1'] server mac:[u'umac-64-etm@openssh.com', u'umac-128-etm@openssh.com', u'hmac-sha2-256-etm@openssh.com', u'hmac-sha2-512-etm@openssh.com', u'hmac-sha1-etm@openssh.com', u'umac-64@openssh.com', u'umac-128@openssh.com', u'hmac-sha2-256', u'hmac-sha2-512', u'hmac-sha1'] client compress:[u'none', u'zlib@openssh.com'] server compress:[u'none', u'zlib@openssh.com'] client lang:[u''] server lang:[u''] kex follows?False
paramiko.transport: DEBUG: Kex agreed: ecdh-sha2-nistp256
paramiko.transport: DEBUG: HostKey agreed: ssh-ed25519
paramiko.transport: DEBUG: Cipher agreed: aes128-ctr
paramiko.transport: DEBUG: MAC agreed: hmac-sha2-256
paramiko.transport: DEBUG: Compression agreed: none
paramiko.transport: DEBUG: kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
paramiko.transport: DEBUG: Switch to new keys ...
paramiko.transport: DEBUG: Adding ssh-ed25519 host key for [192.168.188.92]:8022: c857a7912437d0104fb5282bbb400087
paramiko.transport: DEBUG: Trying discovered key 6fdbaea68289971a2e37c6caca1c87f6 in /home/infrasim/.ssh/id_rsa
paramiko.transport: DEBUG: userauth is OK
paramiko.transport: INFO: Authentication (publickey) failed.
paramiko.transport: DEBUG: userauth is OK
paramiko.transport: INFO: Authentication (password) successful!
--------------------- >> end captured logging << --------------------- 
<br> 
<br></p></details>

File: infrasim/chassis/emu_data.py [hunk 1]
```diff
@@ -179,8 +184,9 @@ def ChangeChassisInfo(self, info):
         # Refer to EMC technical white paper in Dell Community, the Type is 0x17
         result.append(0x17)
         self.__fill_table(result, _ori_values)
-
-        self._data_area[FruCmd.CHASSIS_INFO_AREA] = {"start": 0, "end": 0, "data": result}
+        if len(result) > len(_data):
```
Comment (XiaowenJiang) [inline infrasim/chassis/emu_data.py]: what will be filled here if the result is too long?

File: test/functional/test_chassis.py [hunk 2]
```diff
@@ -103,6 +103,7 @@ def start_chassis():
     global ssh2
     global chassis
     conf = fixtures.ChassisConfig().get_chassis_info()
+    conf["type"] = "s2600kp"
```
Comment (XiaowenJiang) [inline test/functional/test_chassis.py]: 👍

File: infrasim/chassis/emu_data.py (hunk 1 above)
Comment (weia5-emc) [reply infrasim/chassis/emu_data.py]: nothing changed if new data is too long.

**Pull request merged**